File: PathoFroh/src/main/java/com/patho/main/service/PDFService.java
package com.patho.main.service;

import com.patho.main.config.PathoConfig;
import com.patho.main.model.PDFContainer;
import com.patho.main.model.interfaces.DataList;
import com.patho.main.model.patient.DiagnosisRevision;
import com.patho.main.model.patient.Patient;
import com.patho.main.model.patient.Task;
import com.patho.main.model.patient.miscellaneous.BioBank;
import com.patho.main.model.patient.miscellaneous.Council;
import com.patho.main.repository.jpa.*;
import com.patho.main.repository.miscellaneous.MediaRepository;
import com.patho.main.template.PrintDocument;
import com.patho.main.template.PrintDocumentType;
import com.patho.main.util.helper.HistoUtil;
import com.patho.main.util.pdf.LazyPDFReturnHandler;
import com.patho.main.util.pdf.PDFCreator;
import lombok.Getter;
import lombok.Setter;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.rendering.ImageType;
import org.apache.pdfbox.rendering.PDFRenderer;
import org.hibernate.Hibernate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

@Service
@Transactional
public class PDFService extends AbstractService {

    @Autowired
    private PatientRepository patientRepository;

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private BioBankRepository bioBankRepository;

    @Autowired
    private MediaRepository mediaRepository;

    @Autowired
    private PDFRepository pdfRepository;

    @Autowired
    private CouncilRepository councilRepository;

    @Autowired
    private PathoConfig pathoConfig;

    /**
     * Creates a new empty pdf with no file on the hard drive. This is a "virtual"
     * pdf.
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PDFInfo pdfInfo) throws FileNotFoundException {
        return createAndAttachPDF(dataList, pdfInfo, null, false);
    }

    /**
     * Creates a new pdf and saves the content to the hard drive. If a target folder
     * is passed via pdfInfo, the targetFolder will be used, otherwise the dataLists
     * targetFolder is used.
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PDFInfo pdfInfo, byte[] content)
            throws FileNotFoundException {
        return createAndAttachPDF(dataList, pdfInfo, content, false);
    }

    /**
     * Creates a new pdf and saves the content to the hard drive. If a target folder
     * is passed via pdfInfo, the targetFolder will be used, otherwise the dataLists
     * targetFolder is used.
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PDFInfo pdfInfo, byte[] content, boolean createThumbnail)
            throws FileNotFoundException {

        PDFContainer pdf = newPDF(pdfInfo,
                pdfInfo.getTargetFolder() != null ? pdfInfo.getTargetFolder() : dataList.getFileRepositoryBase());

        pdf = pdfRepository.save(pdf);
        dataList = saveORAttachPDF(dataList, pdf);

        try {
            if (content != null) {
                logger.debug("Saving pdf to disk");
                mediaRepository.saveBytes(content, pdf.getPath());
                if (createThumbnail) {
                    logger.debug("Creating thumbnail");
                    mediaRepository.saveImage(
                            gerateThumbnail(content, 0, pathoConfig.getFileSettings().getThumbnailDPI()),
                            pdf.getThumbnail());
                }
            } else
                logger.debug("Virtual PDF no content!");
        } catch (Exception e) {
            e.printStackTrace();
            logger.debug("Error while saving file, removing file from datalists");
            dataList = deletePdf(dataList, pdf);
            throw new IllegalAccessError("Could not updaload data");
        }

        return new PDFReturn(dataList, pdf);
    }

    /**
     * Creates a new PDF Object an runs the PDFCreator in order to create a pdf from
     * the printDocument
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PrintDocument printDocument) throws FileNotFoundException {
        return createAndAttachPDF(dataList, printDocument, new PDFInfo(printDocument.getGeneratedFileName(),
                printDocument.getDocumentType(), "", "", dataList.getFileRepositoryBase()), false, false, null);
    }

    /**
     * Creates a new PDF Object an runs the PDFCreator in order to create a pdf from
     * the printDocument
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PrintDocument printDocument, boolean createThumbnail)
            throws FileNotFoundException {
        return createAndAttachPDF(
                dataList, printDocument, new PDFInfo(printDocument.getGeneratedFileName(),
                        printDocument.getDocumentType(), "", "", dataList.getFileRepositoryBase()),
                createThumbnail, false, null);
    }

    /**
     * Creates a new PDF Object an runs the PDFCreator in order to create a pdf from
     * the printDocument. For PDF Object creation data from the pdfInfo objects are
     * used.
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PrintDocument printDocument, PDFInfo pdfInfo)
            throws FileNotFoundException {
        return createAndAttachPDF(dataList, printDocument, pdfInfo, false, false, null);
    }

    /**
     * Creates a new PDF Object an runs the PDFCreator in order to create a pdf from
     * the printDocument. For PDF Object creation data from the pdfInfo objects are
     * used. If nonBlocking is true the new PDf Obejct will be returnd and the pdf
     * will be generated in a new Thread.
     */
    public PDFReturn createAndAttachPDF(DataList dataList, PrintDocument printDocument, PDFInfo pdfInfo,
                                        boolean createThumbnail, boolean nonBlocking, LazyPDFReturnHandler returnHandler)
            throws FileNotFoundException {
        PDFContainer pdf = newPDF(pdfInfo,
                pdfInfo.getTargetFolder() != null ? pdfInfo.getTargetFolder() : dataList.getFileRepositoryBase());

        pdf = pdfRepository.save(pdf);
        dataList = saveORAttachPDF(dataList, pdf);

        try {
            if (!nonBlocking)
                pdf = new PDFCreator().updateExistingPDF(printDocument, pdf, createThumbnail);
            else {
                new PDFCreator().updateExistingPDFNonBlocking(printDocument, pdf, createThumbnail,
                        returnHandler != null ? returnHandler : new LazyPDFReturnHandler() {
                            @Override
                            public void returnPDFContent(PDFContainer container, String uuid) {
                                logger.debug("Returning PDF Creation withou result");
                            }
                        });
            }
        } catch (Exception e) {
            e.printStackTrace();
            logger.debug("Error while saving file, removing file from datalists");
            // do not delete
            // TODO handle erro
            // dataList = deletePdf(dataList, pdf);
            throw new IllegalAccessError("Could not updaload data");
        }

        return new PDFReturn(dataList, pdf);
    }

    /**
     * Updates a already created PDF Object. If nonBlocking is true, the pdf
     * creation will be done in a seperate thread.
     */
    public PDFContainer updateAttachedPDF(PDFContainer pdfContainer, PrintDocument printDocument,
                                          boolean createThumbnail, boolean nonBlocking) {
        if (!nonBlocking)
            pdfContainer = new PDFCreator().updateExistingPDF(printDocument, pdfContainer, createThumbnail);
        else {
            new PDFCreator().updateExistingPDFNonBlocking(printDocument, pdfContainer, createThumbnail,
                    new LazyPDFReturnHandler() {
                        @Override
                        public void returnPDFContent(PDFContainer container, String uuid) {
                        }
                    });
        }
        return pdfContainer;
    }

    /**
     * Removes a pdf form the datalist and saves the change.
     */
    public DataList removePdf(DataList dataList, PDFContainer pdfContainer) {
        dataList.removeReport(pdfContainer);
        return saveDataList(dataList, resourceBundle.get("log.pdf.removed", pdfContainer.getName()));
    }

    /**
     * Removes a pdf form the datalist and deletes the content from the hard drive.
     * Als this method call will save all changes to the database.
     */
    public DataList deletePdf(DataList dataList, PDFContainer pdfContainer) {

        dataList = removePdf(dataList, pdfContainer);

        if (pdfContainer.getPath() != null)
            mediaRepository.delete(pdfContainer.getPath());
        if (pdfContainer.getThumbnail() != null)
            mediaRepository.delete(pdfContainer.getThumbnail());

        pdfRepository.delete(pdfContainer);

        return dataList;
    }

    /**
     * Moves a pdf from one datalist to the target datalist. Notice the pdf storen
     * on the hard disk will not be moved if the targetDatalist has an other folder.
     */
    public void movePdf(DataList from, DataList to, PDFContainer pdfContainer) {
        removePdf(from, pdfContainer);
        saveORAttachPDF(to, pdfContainer);
    }

    public void movePdf(List<DataList> from, DataList to, PDFContainer pdfContainer) {
        for (DataList dataList : from) {
            if (dataList.containsReport(pdfContainer)) {
                if (dataList.equals(to))
                    return;
                dataList = removePdf(dataList, pdfContainer);
                break;
            }
        }

        saveORAttachPDF(to, pdfContainer);
    }

    public void copyPdf() {

    }

    /**
     * Adds a pdf to the datalist and saves the datalist depending on the datalists
     * type.
     */
    private DataList saveORAttachPDF(DataList dataList, PDFContainer pdfContainer) {
        dataList.getAttachedPdfs().add(pdfContainer);
        return saveDataList(dataList, resourceBundle.get("log.pdf.uploaded", pdfContainer.getName()));
    }

    /**
     * Saves the datalist depending on the dtatalists type.
     */
    public DataList saveDataList(DataList dataList, String log) {
        if (dataList instanceof Patient) {
            return patientRepository.save((Patient) dataList, log);
        } else if (dataList instanceof Task) {
            return taskRepository.save((Task) dataList, log);
        } else if (dataList instanceof BioBank) {
            return bioBankRepository.save((BioBank) dataList, log);
        } else if (dataList instanceof Council) {
            return councilRepository.save((Council) dataList, log);
        } else {
            throw new IllegalArgumentException("List type not supported (" + dataList + ")");
        }
    }

    /**
     * Creates a new PDF Object, does not save PDF Data to database.
     */
    private PDFContainer newPDF(PDFInfo pdfInfo, File targetFolder) throws FileNotFoundException {
        return newPDF(pdfInfo.getName(), pdfInfo.getDocumentType(), pdfInfo.getCommentary(),
                pdfInfo.getCommentaryIntern(), targetFolder);
    }

    /**
     * Creates a new PDF Object, does not save PDF Data to database.
     */
    private PDFContainer newPDF(String name, PrintDocumentType documentType, String commentary, String commentaryIntern,
                                File targetFolder) throws FileNotFoundException {
        PDFContainer pdf = new PDFContainer();
        pdf.setName(name);
        pdf.setType(documentType);
        pdf.setCommentary(commentary);
        pdf.setIntern(commentaryIntern);
        getUniquePDF(targetFolder, pdf);
        return pdf;
    }

    public Patient initializeDataListTree(Patient patient) {
        patient = patientRepository.save(patient);

        Hibernate.initialize(patient.getAttachedPdfs());
        for (Task task : patient.getTasks()) {
            Hibernate.initialize(task.getAttachedPdfs());
            Hibernate.initialize(task.getCouncils());

            for (Council council : task.getCouncils()) {
                Hibernate.initialize(council.getAttachedPdfs());
            }
        }

        return patient;
    }

    public File createThumbnail(File output, byte[] pdf) {
        return createThumbnail(output, pdf, 0, pathoConfig.getFileSettings().getThumbnailDPI());
    }

    public File createThumbnail(File output, File input, int pageNo, int dpi) {
        byte[] load = mediaRepository.getBytes(input.getPath());
        return createThumbnail(output, load, pageNo, dpi);
    }

    public File createThumbnail(File output, byte[] pdf, int pageNo, int dpi) {
        mediaRepository.saveImage(gerateThumbnail(pdf, 0, pathoConfig.getFileSettings().getThumbnailDPI()), output);
        return output;
    }

    public BufferedImage gerateThumbnail(File path, int pageNo, int dpi) {
        return gerateThumbnail(mediaRepository.getBytes(path.getPath()), pageNo, dpi);
    }

    /**
     * Returns a buffered Image from a pdf page, page number is zero based
     */
    public BufferedImage gerateThumbnail(byte[] pdf, int pageNo, int dpi) {
        PDDocument document = null;
        try {
            document = PDDocument.load(pdf);

            PDFRenderer pdfRenderer = new PDFRenderer(document);

            if (pageNo < document.getNumberOfPages())
                return pdfRenderer.renderImageWithDPI(pageNo, dpi, ImageType.RGB);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (document != null)
                try {
                    document.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
        }
        return null;
    }

    public List<DataList> getDataListsOfPatient(Patient p) {
        List<DataList> result = new ArrayList<DataList>();

        result.add(p);

        for (Task task : p.getTasks()) {
            result.add(task);

            Optional<BioBank> bio = bioBankRepository.findOptionalByTask(task);

            if (bio.isPresent())
                result.add(bio.get());

            for (Council c : task.getCouncils()) {
                result.add(c);
            }
        }

        return result;
    }

    /**
     * Gets a uinique file within the folder
     */
    public PDFContainer getUniquePDF(File folder) throws FileNotFoundException {
        return getUniquePDF(folder, new PDFContainer());
    }

    /**
     * Gets a uinique file within the folder
     */
    public PDFContainer getUniquePDF(File folder, PDFContainer container) throws FileNotFoundException {
        File uniqueFile = new File(folder, mediaRepository.getUniqueName(folder, ".pdf"));
        container.setPath(uniqueFile.getPath());
        container.setThumbnail(uniqueFile.getPath().replace(".pdf", ".png"));
        return container;
    }

    /**
     * Search for the parentdatalist of the given pdf.
     */
    public static DataList getParentOfPDF(List<DataList> dataLists, PDFContainer container) {

        for (DataList dataList : dataLists) {
            for (PDFContainer pContainer : dataList.getAttachedPdfs()) {
                if (pContainer.equals(container))
                    return dataList;
            }
        }

        return null;
    }

    /**
     * Searches for a datalist within all datalists e.g. of a patient. This is to
     * find the datalist with the current version.
     */
    public static DataList getDatalistFromDatalists(List<DataList> dataLists, DataList findeThisList) {
        for (DataList dataList : dataLists) {
            if (dataList.getId() == findeThisList.getId() && dataList.getClass().equals(findeThisList.getClass()))
                return dataList;
        }

        return null;
    }

    /**
     * Returns a list of pdfs with the given type.
     */
    public static List<PDFContainer> findPDFsByDocumentType(Set<PDFContainer> containers, PrintDocumentType type) {
        return containers.stream().filter(p -> p.getType().equals(type)).collect(Collectors.toList());
    }

    /**
     * Returns the pdf with the most recent creation date.
     */
    public static PDFContainer findNewestPDF(Set<PDFContainer> containers) {
        if (containers.size() == 0)
            return null;

        PDFContainer latest = HistoUtil.getFirstElement(containers);

        for (PDFContainer pdfContainer : containers) {
            if (latest.getAudit().getCreatedOn() < pdfContainer.getAudit().getCreatedOn())
                latest = pdfContainer;
        }

        return latest;
    }

    /**
     * Sorts an set an retuns it as soreted list. (Sorted by creation date)
     */
    public static List<PDFContainer> sortPDFsByDate(Set<PDFContainer> set, boolean asc) {
        List<PDFContainer> list = new ArrayList<PDFContainer>(set);

        // sorting
        Collections.sort(list, (PDFContainer p1, PDFContainer p2) -> {
            if (p1.getAudit().getCreatedOn() == p2.getAudit().getCreatedOn()) {
                return 0;
            } else if (p1.getAudit().getCreatedOn() < p2.getAudit().getCreatedOn()) {
                return asc ? -1 : 1;
            } else {
                return asc ? 1 : -1;
            }
        });

        return list;
    }

    /**
     * Searches for a matching reportIntent report by internal id.
     *
     * @param task
     * @param diagnosisRevision
     * @return
     */
    public static Optional<PDFContainer> findDiagnosisReport(Task task, DiagnosisRevision diagnosisRevision) {
        String matcher = PDFContainer.MARKER_DIAGNOSIS.replace("$id", String.valueOf(diagnosisRevision.getId()));
        for (PDFContainer container : task.getAttachedPdfs()) {
            if (container.getIntern() != null && container.getIntern().matches(matcher)) {
                return Optional.ofNullable(container);
            }
        }
        return Optional.empty();
    }

    /**
     * Return class for returning the new pdf an the datalist
     *
     * @author dvk-glatza
     */
    public static class PDFReturn {
        DataList dataList;
        PDFContainer container;

        public PDFReturn(DataList dataList, PDFContainer container) {
            this.dataList = dataList;
            this.container = container;
        }

        public DataList getDataList() {
            return this.dataList;
        }

        public PDFContainer getContainer() {
            return this.container;
        }

        public void setDataList(DataList dataList) {
            this.dataList = dataList;
        }

        public void setContainer(PDFContainer container) {
            this.container = container;
        }
    }

    /**
     * Helper class for creating or updating pdfs.
     *
     * @author dvk-glatza
     */
    @Getter
    @Setter
    public static class PDFInfo {
        private String name;
        private PrintDocumentType documentType;
        private String commentary;
        private String commentaryIntern;
        private File targetFolder;

        public PDFInfo(String name, PrintDocumentType documentType) {
            this(name, documentType, "", "", null);
        }

        public PDFInfo(String name, PrintDocumentType documentType, File targetFolder) {
            this(name, documentType, "", "", targetFolder);
        }

        public PDFInfo(String name, PrintDocumentType documentType, String commentary, String commentaryIntern) {
            this(name, documentType, commentary, commentaryIntern, null);
        }

        public PDFInfo(String name, PrintDocumentType documentType, String commentary, String commentaryIntern,
                       File targetFolder) {
            this.name = name;
            this.documentType = documentType;
            this.commentary = commentary;
            this.commentaryIntern = commentaryIntern;
            this.targetFolder = targetFolder;
        }
    }
}

File: PathoFroh/src/main/kotlin/com/patho/main/service/PDFService.kt
package com.patho.main.service

import com.patho.main.model.PDFContainer
import com.patho.main.model.interfaces.DataList
import com.patho.main.model.patient.DiagnosisRevision
import com.patho.main.model.patient.Patient
import com.patho.main.model.patient.Task
import com.patho.main.model.patient.miscellaneous.BioBank
import com.patho.main.model.patient.miscellaneous.Council
import com.patho.main.repository.jpa.*
import com.patho.main.repository.miscellaneous.MediaRepository
import com.patho.main.service.impl.SpringContextBridge
import com.patho.main.template.PrintDocument
import com.patho.main.template.PrintDocumentType
import com.patho.main.util.pdf.LazyPDFReturnHandler
import com.patho.main.util.pdf.creator.PDFUpdater
import com.patho.main.util.pdf.creator.PDFUpdaterNonBlocking
import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.rendering.ImageType
import org.apache.pdfbox.rendering.PDFRenderer
import org.hibernate.Hibernate
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.awt.image.BufferedImage
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.util.*
import javax.imageio.ImageIO


@Service()
open class PDFService @Autowired constructor(
        private val patientRepository: PatientRepository,
        private val taskRepository: TaskRepository,
        private val bioBankRepository: BioBankRepository,
        private val councilRepository: CouncilRepository,
        private val pdfRepository: PDFRepository,
        private val mediaRepository: MediaRepository) : AbstractService() {

    companion object {
        /**
         * Search a list of datalist for the list containing the pdf. Returns null if not found
         */
        @JvmStatic
        fun getParentDatalistOfPDF(dataLists: List<DataList>, container: PDFContainer): DataList? {
            for (dataList in dataLists) {
                for (pContainer in dataList.attachedPdfs) {
                    if (pContainer.equals(container)) return dataList
                }
            }
            return null
        }

        /**
         * Searches for a pdf matching the diagnosis number, determine by internal id
         */
        @JvmStatic
        fun findDiagnosisReport(task: Task, diagnosisRevision: DiagnosisRevision): PDFContainer? {
            val matcher = PDFContainer.MARKER_DIAGNOSIS.replace("\$id", diagnosisRevision.id.toString())
            for (container in task.attachedPdfs) {
                if (container.intern.matches(Regex(matcher))) {
                    return container
                }
            }
            return null
        }

        /**
         * Search for a datalist in list of datalists.This is to find the datalist with the current version.
         */
        @JvmStatic
        fun findDataList(dataLists: List<DataList>, find: DataList): DataList? {
            return dataLists.firstOrNull { it.id == find.id && it.javaClass == find.javaClass }
        }

        /**
         * Returns a flat list of all datalists of a patient
         */
        @JvmStatic
        fun flatListOfDataLists(p: Patient): List<DataList>? {
            val result: MutableList<DataList> = ArrayList()
            result.add(p)
            p.tasks.forEach { t ->
                result.add(t)

                val bio: Optional<BioBank> = SpringContextBridge.services().bioBankRepository.findOptionalByTask(t)
                if (bio.isPresent) result.add(bio.get())

                t.councils.forEach { c ->
                    result.add(c)
                }
            }
            return result
        }
    }


    /**
     *  Creates a new PDF and adds it to a datalist. Stores the list in the database and the pdf data to the disk
     */
    @Transactional
    open fun createPDFAndAddToDataList(dataList: DataList, content: ByteArray?, createThumbnail: Boolean, name: String, type: PrintDocumentType, commentary: String = "", commentaryIntern: String = "", targetDirectory: String = dataList.fileRepositoryBase.path): PDFReturn {
        var resultPDF = getSavedUniquePDF(targetDirectory, createThumbnail, name, type, commentary, commentaryIntern)

        var dataListResult = saveAndAttachPDFToList(dataList, resultPDF)

        try {
            if (content != null) {
                logger.debug("Saving pdf to disk")
                mediaRepository.saveBytes(content, resultPDF.path)

                if (createThumbnail)
                    generateAndSaveThumbnail(content, resultPDF.thumbnail)
            }
        } catch (e: Exception) {
            logger.debug("Error while saving file, removing file from datalists")
            e.printStackTrace()
            dataListResult = removeAndDeletePDF(dataListResult, resultPDF)
        }

        return PDFReturn(dataListResult, resultPDF)
    }


    /**
     *  Creates a new PDF and adds it to a datalist. Stores the list in the database and the pdf data to the disk
     */
    @Transactional
    open fun createPDFAndAddToDataList(dataList: DataList, printDocument: PrintDocument, createThumbnail: Boolean, name: String = printDocument.generatedFileName, type: PrintDocumentType = printDocument.documentType, commentary: String = "", commentaryIntern: String = "", targetDirectory: String = dataList.fileRepositoryBase.path, nonBlocking: Boolean = false, returnHandler: LazyPDFReturnHandler? = null): PDFReturn {
        var resultPDF = getSavedUniquePDF(targetDirectory, createThumbnail, name, type, commentary, commentaryIntern)

        resultPDF = pdfRepository.save(resultPDF)

        val dataListResult = saveAndAttachPDFToList(dataList, resultPDF)

        if (!nonBlocking)
            resultPDF = PDFUpdater().update(printDocument, resultPDF, createThumbnail)
        else
            PDFUpdaterNonBlocking().update(printDocument, resultPDF, createThumbnail, returnHandler
                    ?: LazyPDFReturnHandler() { _: PDFContainer, _: String -> })

        return PDFReturn(dataListResult, resultPDF)
    }

    /**
     * Updates an existing pdf. Blocks the program
     */
    @Transactional
    open fun updatePDF(container: PDFContainer, printDocument: PrintDocument, createThumbnail: Boolean, nonBlocking: Boolean = false, returnHandler: LazyPDFReturnHandler? = null) {
        if (!nonBlocking)
            PDFUpdater().update(printDocument, container, createThumbnail)
        else
            PDFUpdaterNonBlocking().update(printDocument, container, createThumbnail, LazyPDFReturnHandler() { _: PDFContainer, _: String -> })
    }

    @Transactional
    open fun removePDF(dataList: DataList, pdfContainer: PDFContainer): DataList {
        dataList.removeReport(pdfContainer)
        return saveDataList(dataList, resourceBundle["log.pdf.removed", pdfContainer.name])
    }

    /**
     * Adds a pdf to a datalist an save the changes in the database
     */
    @Transactional
    open fun saveAndAttachPDFToList(dataList: DataList, pdfContainer: PDFContainer): DataList {
        if (!dataList.attachedPdfs.contains(pdfContainer)) {
            dataList.attachedPdfs.add(pdfContainer)
        }
        return saveDataList(dataList, resourceBundle.get("log.pdf.uploaded", pdfContainer.name))
    }

    /**
     * Saves all know types of dataLists
     */
    @Transactional
    open fun saveDataList(dataList: DataList, log: String): DataList {
        return when (dataList) {
            is Patient -> {
                patientRepository.save(dataList as Patient, log) as DataList
            }
            is Task -> {
                taskRepository.save(dataList as Task, log) as DataList
            }
            is BioBank -> {
                bioBankRepository.save(dataList as BioBank, log) as DataList
            }
            is Council -> {
                councilRepository.save(dataList as Council, log) as DataList
            }
            else -> {
                throw IllegalArgumentException("List type not supported ($dataList)")
            }
        }
    }

    /**
     * Returns a pdf with a unique name within the target dir, does not save content to disk. Sets passed variables. Saves
     * pdfContainer in database
     */
    @Transactional
    open fun getSavedUniquePDF(targetDirectory: String, createThumbnail: Boolean, name: String, type: PrintDocumentType, commentary: String, commentaryIntern: String): PDFContainer {
        val resultPDF = getUniquePDF(targetDirectory, createThumbnail, name, type, commentary, commentaryIntern)
        return pdfRepository.save(resultPDF)
    }

    /**
     * Returns a pdf with a unique name within the target dir, does not save to disk. Sets passed variables.
     */
    @Transactional
    open fun getUniquePDF(targetDirectory: String, createThumbnail: Boolean, name: String, type: PrintDocumentType, commentary: String, commentaryIntern: String): PDFContainer {
        val pdf = getUniquePDF(targetDirectory, createThumbnail)
        pdf.name = name
        pdf.type = type
        pdf.commentary = commentary
        pdf.intern = commentaryIntern
        return pdf
    }

    /**
     * Returns a pdf with a unique name within the target dir, does not save to disk
     */
    @Transactional
    open fun getUniquePDF(targetDirectory: String, createThumbnail: Boolean): PDFContainer {
        val outputName = SpringContextBridge.services().mediaRepository.getUniqueName(targetDirectory, ".pdf")
        val outputFile = File(targetDirectory, outputName)
        val outputIMG = File(targetDirectory, outputName.replace(".pdf", ".png"))

        logger.debug("PDF with output file ${outputFile.path}; dir ${SpringContextBridge.services().mediaRepository.getFileForPath(targetDirectory).absolutePath}")

        val pdf = PDFContainer()
        pdf.path = outputFile.path
        pdf.thumbnail = if (createThumbnail) outputIMG.path else ""

        return pdf
    }

    /**
     * Removes a pdf from the list. Changes are also store in database
     */
    @Transactional
    open fun removePDFt(dataList: DataList, pdfContainer: PDFContainer): DataList {
        dataList.removeReport(pdfContainer)
        return saveDataList(dataList, resourceBundle["log.pdf.removed", pdfContainer.name])
    }

    /**
     * Removes a pdf from the list an deletes it content from the disk. Changes are also store in database
     */
    @Transactional
    open fun removeAndDeletePDF(dataList: DataList, pdfContainer: PDFContainer): DataList {
        val result = removePDFt(dataList, pdfContainer)
        mediaRepository.delete(pdfContainer.path)
        mediaRepository.delete(pdfContainer.thumbnail)
        return dataList
    }

    /**
     * Moves a pdf from one datalist to another
     */
    @Transactional
    open fun movePDF(source: DataList, target: DataList, pdfContainer: PDFContainer) {
        removePDF(source, pdfContainer)
        saveAndAttachPDFToList(target, pdfContainer)
    }

    /**
     * Search a list of dataList for the parent of the pdf. If found the pdf will be moved to the target list
     */
    @Transactional
    open fun movePDF(source: List<DataList>, target: DataList, pdfContainer: PDFContainer) {
        val sourceList = source.firstOrNull { it.containsReport(pdfContainer) }
        if (sourceList == target || sourceList == null)
            return

        movePDF(sourceList, target, pdfContainer)
    }

    /**
     * Generates a thumbnail from a byteArray (pdf content). Reads source pdf from disk. Saves the result to the target file.
     */
    @Transactional
    open fun generateAndSaveThumbnail(source: String, target: String): Boolean {
        return generateAndSaveThumbnail(source, target, 0, SpringContextBridge.services().pathoConfig.fileSettings.thumbnailDPI)
    }

    /**
     * Generates a thumbnail from a byteArray (pdf content). Reads source pdf from disk. Saves the result to the target file.
     */
    @Transactional
    open fun generateAndSaveThumbnail(source: String, target: String, pageNo: Int, dpi: Int): Boolean {
        val img = generateThumbnail(source, pageNo, dpi) ?: return false
        mediaRepository.saveImage(img, target);
        return true
    }

    /**
     * Generates a thumbnail from a byteArray (pdf content). Saves the result to the target file.
     */
    @Transactional
    open fun generateAndSaveThumbnail(source: ByteArray, target: String): Boolean {
        return generateAndSaveThumbnail(source, target, 0, SpringContextBridge.services().pathoConfig.fileSettings.thumbnailDPI)
    }

    /**
     * Generates a thumbnail from a byteArray (pdf content). Saves the result to the target file.
     */
    @Transactional
    open fun generateAndSaveThumbnail(source: ByteArray, target: String, pageNo: Int, dpi: Int): Boolean {
        val img = generateThumbnail(source, pageNo, dpi) ?: return false
        mediaRepository.saveImage(img, target);
        return true
    }

    /**
     * Generates a thumbnail from a byteArry (pdf content). Reads pdf from disk
     */
    @Transactional
    open fun generateThumbnail(path: String, pageNo: Int, dpi: Int): BufferedImage? {
        return generateThumbnail(mediaRepository.getBytes(path), pageNo, dpi)
    }

    /**
     * Generates a thumbnail from a byteArry (pdf content). Reads pdf from disk
     */
    @Transactional
    open fun generateThumbnail(content: ByteArray, pageNo: Int): BufferedImage? {
        return generateThumbnail(content, pageNo, SpringContextBridge.services().pathoConfig.fileSettings.thumbnailDPI)
    }

    /**
     * Generates a thumbnail from a byteArray (pdf content)
     */
    @Transactional
    open fun generateThumbnail(content: ByteArray, pageNo: Int, dpi: Int): BufferedImage? {
        var document: PDDocument? = null
        try {
            document = PDDocument.load(content)
            val pdfRenderer = PDFRenderer(document)
            if (pageNo < document!!.numberOfPages) return pdfRenderer.renderImageWithDPI(pageNo, dpi.toFloat(), ImageType.RGB)
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (document != null) try {
                document.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        return null
    }

    /**
     * Returns an images as an bytearray
     */
    open fun thumbnailToByteArray(image: BufferedImage): ByteArray {
        val baos = ByteArrayOutputStream()
        ImageIO.write(image, "png", baos)
        baos.flush()
        return baos.toByteArray()
    }

    /**
     * Initializes all datalists of the patient
     */
    @Transactional
    open fun initializeDataListTree(p: Patient): Patient {
        val patient = patientRepository.save(p)

        Hibernate.initialize(patient.attachedPdfs)
        for (task in patient.tasks) {
            Hibernate.initialize(task.attachedPdfs)
            Hibernate.initialize(task.councils)
            for (council in task.councils) {
                Hibernate.initialize(council.attachedPdfs)
            }
        }
        return patient
    }

    class PDFReturn(val dataList: DataList, val container: PDFContainer)
}